Source/WebKit/Shared/JavaScriptEvaluationResult.mm

```diff
@@ -26,13 +26,225 @@
 #import "config.h"
 #import "JavaScriptEvaluationResult.h"
 
-#import "APISerializedScriptValue.h"
+#include "APIArray.h"
+#include "APISerializedScriptValue.h"
+#include "APIString.h"
+#include "CoreIPCNumber.h"
+#include "WKAPICast.h"
+#include "WKMutableArray.h"
+#include "WKMutableDictionary.h"
+#include "WKNumber.h"
+#include "WKString.h"
+#include <JavaScriptCore/JSContext.h>
+#include <JavaScriptCore/JSValue.h>
+#include <WebCore/ExceptionDetails.h>
+#include <WebCore/SerializedScriptValue.h>
+#include <wtf/RunLoop.h>
 
 namespace WebKit {
 
+JavaScriptEvaluationResult::JavaScriptEvaluationResult(JSObjectID root, HashMap<JSObjectID, Variant>&& map)
+    : m_map(WTFMove(map))
+    , m_root(root) { }
+
+RetainPtr<id> JavaScriptEvaluationResult::toID(const Variant& root) const
+{
+    return WTF::switchOn(root, [] (NullType nullType) -> RetainPtr<id> {
+        switch (nullType) {
+        case NullType::NSNull:
+            return NSNull.null;
+        case NullType::NullPointer:
+            break;
+        }
+        return nullptr;
+    }, [] (bool value) -> RetainPtr<id> {
+        return value ? @YES : @NO;
+    }, [] (const CoreIPCNumber& value) -> RetainPtr<id> {
+        return value.toID();
+    }, [] (const String& value) -> RetainPtr<id> {
+        return (NSString *)value;
+    }, [] (Seconds value) -> RetainPtr<id> {
+        return [NSDate dateWithTimeIntervalSince1970:value.seconds()];
+    }, [&] (const Vector<JSObjectID>& vector) -> RetainPtr<id> {
+        RetainPtr result = adoptNS([[NSMutableArray alloc] initWithCapacity:vector.size()]);
+        for (auto objectID : vector) {
+            if (RetainPtr element = toID(objectID))
+                [result addObject:element.get()];
+        }
+        return result;
+    }, [&] (const HashMap<JSObjectID, JSObjectID>& map) -> RetainPtr<id> {
+        RetainPtr result = adoptNS([[NSMutableDictionary alloc] initWithCapacity:map.size()]);
+        for (auto [keyID, valueID] : map) {
+            RetainPtr key = toID(keyID);
+            if (!key)
+                continue;
+            RetainPtr value = toID(valueID);
+            if (!value)
+                continue;
+            [result setObject:value.get() forKey:key.get()];
+        }
+        return result;
+    });
+}
+
+RetainPtr<id> JavaScriptEvaluationResult::toID(JSObjectID root) const
+{
+    if (RetainPtr object = m_instantiatedObjects.get(root))
+        return object;
+
+    auto it = m_map.find(root);
+    if (it == m_map.end())
+        return nil;
+
+    RetainPtr result = toID(it->value);
+    m_instantiatedObjects.set(root, result);
```

Comment:
We should update m_instantiatedObjects with before recursively converting its children, to avoid invoking toID() for the same ID.